hash-table/chaining.js

```diff
@@ -0,0 +1,147 @@
+const { Node, DoublyLinkedList } = require('../linked-list/doubly-linked-list')
+
+const DEFAULT_CAPACITY = 8
+class ChainingHashTable {
+  constructor() {
+    this.arr = new Array(DEFAULT_CAPACITY)
+    this.capacity = DEFAULT_CAPACITY
+    this.size = 0
+  }
+
+  insert(value, key) {
+    const index = this.hash(key)
+    if (!this.arr[index]) {
+      const linkList = new DoublyLinkedList()
+      linkList.pushFront(value, key)
+      this.arr[index] = linkList
+      this.size++
+      if (this.size === this.capacity) {
+        this.grow()
+      }
+      return
+    }
+
+    const isExisting = this.search(key)
+    if (!isExisting) {
+      this.arr[index].pushBack(value, key)
+      return
+    }
+    let cur = this.arr[index].tail
+    while (cur) {
+      if (cur.key === key) {
+        cur.value = value
+        break
+      }
+      cur = cur.prev
+    }
+  }
+
+  delete(key) {
+    const index = this.hash(key)
+    
+    if (!this.arr[index]) return
+    if (this.arr[index].length === 1) {
+      this.arr[index] = undefined
+      this.size--
+      if (this.capacity > DEFAULT_CAPACITY && this.size <= Math.floor(this.capacity / 4)) {
+        this.shrink()
+      }
+      return
+    }
+    
+    let cur = this.arr[index].tail
+    let count = 0
+    while (cur) {
+      if (cur.key === key) {
+        let length = this.arr[index].length
+        this.arr[index].removeAt(length - count - 1)
+        break
+      }
+      count++
+      cur = cur.prev
+    } 
+  }
+
+  search(key) {
+    if (!this.size) return null
+    const index = this.hash(key)
+    
+    if (!this.arr[index]) {
+      return null
+    }
+
+    let cur = this.arr[index].tail
+    while (cur) {
+      if (cur.key === key) {
+        return cur.value
+      }
+      cur = cur.prev
+    }
+  }
+
+  // private methods
+  hash(key) {
```

Comment:
A common convention I see for marking private properties/methods in JS is to add `_` as prefix. If you know any cooler tricks, tell me. :)